File: backend/app/services/weather_service.py
import httpx
from typing import Dict, Any
from app.config import settings

class WeatherService:
    @staticmethod
    async def get_weather(city: str) -> Dict[str, Any]:
        """Fetch weather from OpenWeatherMap if key is provided, else return realistic mock data."""
        if settings.WEATHER_API_KEY:
            try:
                async with httpx.AsyncClient(timeout=5.0) as client:
                    url = f"https://api.openweathermap.org/data/2.5/weather?q={city}&appid={settings.WEATHER_API_KEY}&units=metric"
                    res = await client.get(url)
                    if res.status_code == 200:
                        data = res.json()
                        temp = data.get("main", {}).get("temp", 25)
                        condition = data.get("weather", [{}])[0].get("main", "Clear")
                        humidity = data.get("main", {}).get("humidity", 60)
                        wind = data.get("wind", {}).get("speed", 12)
                        return {
                            "city": city,
                            "temperature": round(temp, 1),
                            "condition": condition,
                            "humidity": humidity,
                            "rain_probability": 40 if "Rain" in condition else 10,
                            "wind_speed_kmh": round(wind * 3.6, 1),
                            "is_mock": False,
                            "activity_advice": WeatherService._generate_advice(condition, temp)
                        }
            except Exception:
                pass

        # Fallback to realistic mock weather based on Indian destinations
        return WeatherService._mock_weather(city)

    @staticmethod
    def _mock_weather(city: str) -> Dict[str, Any]:
        c_lower = city.lower()
        if "araku" in c_lower:
            temp = 22.0
            condition = "Partly Cloudy"
            rain = 15
            humidity = 62
            wind = 12.0
        elif "vizag" in c_lower or "visakhapatnam" in c_lower:
            temp = 28.5
            condition = "Breezy Coastal"
            rain = 10
            humidity = 72
            wind = 22.0
        elif "goa" in c_lower:
            temp = 30.0
            condition = "Sunny"
            rain = 5
            humidity = 68
            wind = 14.0
        elif "kerala" in c_lower or "munnar" in c_lower:
            temp = 21.0
            condition = "Passing Showers"
            rain = 55
            humidity = 82
            wind = 16.0
        elif "kashmir" in c_lower:
            temp = 14.0
            condition = "Crisp & Clear"
            rain = 5
            humidity = 48
            wind = 9.0
        else:
            temp = 26.0
            condition = "Pleasant"
            rain = 10
            humidity = 58
            wind = 11.0

        return {
            "city": city,
            "temperature": temp,
            "condition": condition,
            "humidity": humidity,
            "rain_probability": rain,
            "wind_speed_kmh": wind,
            "is_mock": True,
            "activity_advice": WeatherService._generate_advice(condition, temp)
        }

    @staticmethod
    def _generate_advice(condition: str, temp: float) -> str:
        if "rain" in condition.lower() or "showers" in condition.lower():
            return "Rain is expected. AI recommends visiting covered museums, heritage tea factories or indoor cultural galleries instead of slippery waterfall trails."
        elif temp > 32:
            return "Warm midday sun forecast. AI recommends scheduling water sports and beach walks before 11:00 AM or after 4:00 PM."
        elif temp < 18:
            return "Cool mountain temperatures. Carry light thermal fleece layers for dawn viewpoints and evening campfires."
        else:
            return "Optimal weather conditions for both outdoor sightseeing and scenic transit."

weather_service = WeatherService()

File: src/services/weatherService.ts
export interface WeatherData {
  city: string;
  temp: number;
  feelsLike: number;
  condition: 'Sunny' | 'Partly Cloudy' | 'Rainy' | 'Breezy' | 'Mist';
  rainProbability: number;
  humidity: number;
  windSpeedKmH: number;
  uvIndex: number;
  airQualityIndex: number;
  forecast: {
    day: string;
    temp: number;
    condition: string;
    rainProb: number;
  }[];
  aiWeatherAdvisory: string;
  recommendedIndoorAlternative?: string;
}

export const weatherService = {
  getWeatherForDestination(destinationName: string): WeatherData {
    const nameLower = destinationName.toLowerCase();

    if (nameLower.includes('araku')) {
      return {
        city: 'Araku Valley',
        temp: 22,
        feelsLike: 21,
        condition: 'Partly Cloudy',
        rainProbability: 15,
        humidity: 62,
        windSpeedKmH: 12,
        uvIndex: 5,
        airQualityIndex: 28, // Pristine mountain air
        forecast: [
          { day: 'Today', temp: 22, condition: 'Partly Cloudy', rainProb: 15 },
          { day: 'Tomorrow', temp: 21, condition: 'Mist & Sunshine', rainProb: 10 },
          { day: 'Day 3', temp: 23, condition: 'Sunny', rainProb: 5 }
        ],
        aiWeatherAdvisory: 'Optimal mountain weather! Morning valley mist clears by 8:30 AM. Perfect light for Borra Caves and Katiki trekking.'
      };
    }

    if (nameLower.includes('vizag') || nameLower.includes('visakhapatnam')) {
      return {
        city: 'Visakhapatnam',
        temp: 28,
        feelsLike: 30,
        condition: 'Breezy',
        rainProbability: 10,
        humidity: 70,
        windSpeedKmH: 22,
        uvIndex: 6,
        airQualityIndex: 45,
        forecast: [
          { day: 'Today', temp: 28, condition: 'Breezy Coastal', rainProb: 10 },
          { day: 'Tomorrow', temp: 29, condition: 'Sunny & Warm', rainProb: 5 },
          { day: 'Day 3', temp: 28, condition: 'Pleasant Winds', rainProb: 15 }
        ],
        aiWeatherAdvisory: 'Fresh sea breezes along Beach Road. Ocean waves are moderate; ideal for evening strolls at Kailasagiri.'
      };
    }

    if (nameLower.includes('goa')) {
      return {
        city: 'Goa',
        temp: 30,
        feelsLike: 33,
        condition: 'Sunny',
        rainProbability: 5,
        humidity: 65,
        windSpeedKmH: 14,
        uvIndex: 8,
        airQualityIndex: 35,
        forecast: [
          { day: 'Today', temp: 30, condition: 'Sunny Beach Weather', rainProb: 5 },
          { day: 'Tomorrow', temp: 31, condition: 'Clear Sky', rainProb: 0 },
          { day: 'Day 3', temp: 30, condition: 'Golden Sunset', rainProb: 5 }
        ],
        aiWeatherAdvisory: 'High UV index at noon. AI recommends scheduling water sports and beach lounging before 11:30 AM or after 3:30 PM.'
      };
    }

    if (nameLower.includes('kerala') || nameLower.includes('munnar')) {
      return {
        city: 'Munnar & Alleppey',
        temp: 21,
        feelsLike: 20,
        condition: 'Rainy',
        rainProbability: 60,
        humidity: 82,
        windSpeedKmH: 16,
        uvIndex: 4,
        airQualityIndex: 22,
        forecast: [
          { day: 'Today', temp: 21, condition: 'Passing Showers', rainProb: 60 },
          { day: 'Tomorrow', temp: 20, condition: 'Misty Rain', rainProb: 55 },
          { day: 'Day 3', temp: 22, condition: 'Clearing Skies', rainProb: 25 }
        ],
        aiWeatherAdvisory: 'Rain is expected tomorrow. AI recommends visiting the Tata Tea Museum & spice processing centers instead of the outdoor waterfall trek.',
        recommendedIndoorAlternative: 'Tata Tea Museum & Cultural Kathakali Hall'
      };
    }

    // Default fallback
    return {
      city: destinationName,
      temp: 26,
      feelsLike: 27,
      condition: 'Sunny',
      rainProbability: 10,
      humidity: 58,
      windSpeedKmH: 10,
      uvIndex: 6,
      airQualityIndex: 40,
      forecast: [
        { day: 'Today', temp: 26, condition: 'Clear', rainProb: 10 },
        { day: 'Tomorrow', temp: 27, condition: 'Sunny', rainProb: 10 },
        { day: 'Day 3', temp: 26, condition: 'Pleasant', rainProb: 15 }
      ],
      aiWeatherAdvisory: 'Favorable travel conditions across outdoor and cultural monuments.'
    };
  }
};
